clippy_lints/src/functions.rs

```diff
@@ -144,35 +226,77 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Functions {
 
+    fn check_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx hir::Item) {
+        let attr = must_use_attr(&item.attrs);
+        if let hir::ItemKind::Fn(ref decl, ref _header, ref _generics, ref body_id) = item.kind {
+            if let Some(attr) = attr {
+                check_needless_must_use(cx, decl, item.hir_id, item.span, attr);
+                return;
+            }
+            if cx.access_levels.is_exported(item.hir_id) {
+                check_must_use_candidate(
+                    cx,
+                    decl,
+                    cx.tcx.hir().body(*body_id),
+                    item.span,
+                    item.hir_id,
+                    "this function could have a `#[must_use]` attribute",
+                );
+            }
+        }
+    }
+
+    fn check_impl_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx hir::ImplItem) {
+        if let hir::ImplItemKind::Method(ref sig, ref body_id) = item.kind {
+            let attr = must_use_attr(&item.attrs);
+            if let Some(attr) = attr {
+                check_needless_must_use(cx, &sig.decl, item.hir_id, item.span, attr);
+            } else if cx.access_levels.is_exported(item.hir_id) {
+                check_must_use_candidate(
+                    cx,
+                    &sig.decl,
+                    cx.tcx.hir().body(*body_id),
+                    item.span,
+                    item.hir_id,
+                    "this method could have a `#[must_use]` attribute",
+                );
+            }
+        }
+    }
+
     fn check_trait_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx hir::TraitItem) {
         if let hir::TraitItemKind::Method(ref sig, ref eid) = item.kind {
             // don't lint extern functions decls, it's not their fault
             if sig.header.abi == Abi::Rust {
                 self.check_arg_number(cx, &sig.decl, item.span);
             }
 
+            let attr = must_use_attr(&item.attrs);
+            if let Some(attr) = attr {
+                check_needless_must_use(cx, &sig.decl, item.hir_id, item.span, attr);
+            }
             if let hir::TraitMethod::Provided(eid) = *eid {
                 let body = cx.tcx.hir().body(eid);
                 self.check_raw_ptr(cx, sig.header.unsafety, &sig.decl, body, item.hir_id);
+
+                if attr.is_none() && cx.access_levels.is_exported(item.hir_id) {
+                    check_must_use_candidate(
+                        cx,
+                        &sig.decl,
+                        body,
+                        item.span,
+                        item.hir_id,
+                        "this method could have a `#[must_use]` attribute",
+                    );
+                }
             }
         }
     }
 }
 
 impl<'a, 'tcx> Functions {
     fn check_arg_number(self, cx: &LateContext<'_, '_>, decl: &hir::FnDecl, span: Span) {
-        // Remove the function body from the span. We can't use `SourceMap::def_span` because the
```

Comment:
It would be nice to keep this comment in the extracted function, too

Reply:
I think I have a better option; I will need some time to apply it though.

Reply:
I removed the function, there's a better solution anyway.